### ISSUE TITLE
PossiblyInvalidArgument reported for array item that was previously filtered with instanceof

### ISSUE THREAD

**Issue by slepic** [vimeo/psalm#4403]
Without changing my code, psalm reported this error after updating vimeo/psalm (3.12.2 => 3.18.2)

https://psalm.dev/r/964e40285c

The error goes away as soos as I change the array_map callback from this:
```
fn ($x) => new Wrapper($x),
```
to
```
fn (InterfaceB $x) => new Wrapper($x),
```

First I thought it was a regression of #3764 
But it is probably just similar

Anyway although just adding typehint to the arrow function parameter fixes the issue, I don't think it should be necesary.

**Comment by jsiefer** [comment]
Well even your filter callback logic is in this example easy to reason about, it could theoretically be more complex. The only thing psalm can reason from your filter callback is that it will return true or false. That information is not enough. Psalm does not run your code it only reasons about your types.

At least that's how I look at it :)

**Comment by slepic** [comment]
I understand your argument, but can you then explain to me why adding the typehint to the map callback, removes the issue without spawning a different error?

If it thinks that the `array_filter` still returns `array<string, TestInstanceA|TestInstanceAB>` why does it not complain about passing that output to `array_map` with a callback accepting `InterfaceB` only?

**Comment by jsiefer** [comment]
Good question,....  you right, it does seem a bit of. It looks like array_filter returns the type ClassA&InterfaceB|ClassAB instead of ClassA|ClassAB.

https://psalm.dev/r/e950619c21

**Comment by slepic** [comment]
But isn't it correct that it returns `(ClassA&InterfaceB)|ClassAB`? That would actually mean that it was able to infer that it only returned those instances of `ClassA|ClassAB` that are also instances of `InterfaceB`, which `ClassAB` is be design and `ClassA` might have a successor class:
```
class ClassC extends ClassA implements InterfaceB {}
```

It might have also inferred the return type as just `ClassAB`, because `ClassA` alone never implemented `InterfaceB` and so it could have been removed from the union.

Nevertheless, the problem must be somewhere up the stream, not in the filter...

**Comment by jsiefer** [comment]
Yes it almost looks like it does parse the function. Though its not a union join, that would have been correct. (ClassA|InterfaceB)|ClassAB

**Comment by muglug** [comment]
Simplified to

```php
<?php

interface A {}
interface B {}

class Wrapper {
    public function __construct(B $input) {}
}

/** @param array<int, A&B> $filtered */
function foo(array $filtered): void {
    \array_map(
        fn ($x) => new Wrapper($x),
        $filtered
    );
}
```

**Comment by muglug** [comment]
Original issue no longer appears, but code above still errors

**Comment by orklah** [comment]
https://psalm.dev/r/a34c12c0ab
